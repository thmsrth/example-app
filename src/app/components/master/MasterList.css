.MasterList {
  width: 100%;
}

.MasterList__icon {
  margin: 0 var(--section-margin);
}

.MasterListHead__description,
.MasterListItem__description {
  text-align: left;
  width: 250px;
}

.MasterList__buttons {
  padding: var(--section-padding);
}

.MasterList__container {
  padding: var(--section-padding);
}

.MasterList__table {
  table-layout: auto;
}

.MasterListItem--active {
  background-color: var(--color-brand-secondary);
}

.MasterListItem {
  cursor: default;
}

.MasterListItemBox__userMap {
  min-width: 14vw;
}
