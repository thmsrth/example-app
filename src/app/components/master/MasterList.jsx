import React from 'react';

import PropTypes from 'prop-types';
import dataItemType from '../types/dataItem';

import { MAIN_PATH } from '../../../../config/constants';

import paginationType from '../../../global/components/Pagination/types/pagination';
import paginationDefaults from '../../../global/components/Pagination/defaults/pagination';
import { historyPush } from '../../../global/utils/UrlHelpers';
import OverviewList from '../../../global/components/OverviewList';

import I18n from '../../utils/I18n';
import SecurityUtil from '../../../global/security/utils/SecurityUtil';

import './MasterList.css';

const MasterList = ({
  data, dataLoading, dataSort, dataPagination,
  fetchSortedItems, updateSortQueryParams,
}) => {
  const keys = [
    {
      name: 'id',
      title: I18n.t('en.data.attribute.id.title'),
      type: 'number',
    },
    {
      name: 'name',
      title: I18n.t('en.data.attribute.name.title'),
      type: 'text',
    },
    {
      name: 'nutritionalInformation',
      title: I18n.t('en.data.attribute.nutritionalInformation.title'),
      type: 'box',
      options: [
        {
          name: 'fat',
          title: `${I18n.t('en.data.attribute.fat.title')}:`,
          type: 'number',
        },
        {
          name: 'protein',
          title: `${I18n.t('en.data.attribute.protein.title')}:`,
          type: 'number',
        },
        {
          name: 'calories',
          title: `${I18n.t('en.data.attribute.calories.title')}:`,
          type: 'number',
        },
      ],
    },
    {
      name: 'status',
      title: I18n.t('en.data.attribute.status.title'),
      type: 'text',
    },
    {
      name: 'createdAt', title: 'Created At', type: 'datetime', format: 'DD/MM/YYYY HH:mm',
    },
    {
      name: 'createdBy', title: 'Created by', type: 'text',
    },
    {
      name: 'updatedAt', title: 'Updated At', type: 'datetime', format: 'DD/MM/YYYY HH:mm',
    },
    {
      name: 'updatedBy', title: 'Updated by', type: 'text',
    },
    {
      name: 'actions',
      title: 'Actions',
      type: 'box',
      options: [
        {
          name: 'detail',
          title: '',
          type: 'action',
          actionParams: ['id'],
          condition: SecurityUtil.isGranted('EXAMPLE_DETAIL'),
          icon: 'search',
        },
      ],
    },
  ];

  if (data && !dataLoading) {
    const actionHandlers = {
      detail: params => historyPush(`${MAIN_PATH}/${params.id}`),
    };
    return (
      <div className="MasterList">
        <div className="MasterList__container">
          {data.length
            ? (
              <OverviewList
                items={data}
                keys={keys}
                handlers={actionHandlers}
                className="MasterList"
                sortedKey={dataSort && dataSort.property}
                sortOrder={dataSort && dataSort.direction}
                fetchSortedItems={fetchSortedItems}
                updateSortQueryParams={updateSortQueryParams}
                serverSidedSorting={dataPagination.totalElements > dataPagination.size}
              />
            )
            : (
              <div className="MasterList__list MasterList__list--empty">
                <h4 className="MasterList__states">
                  {I18n.t('en.data.states.empty')}
                </h4>
              </div>
            )
          }
        </div>
      </div>
    );
  }
  return null;
};

export default MasterList;

MasterList.propTypes = {
  data: PropTypes.arrayOf(dataItemType),
  dataSort: PropTypes.shape({
    property: PropTypes.string,
    direction: PropTypes.string,
  }),
  dataPagination: paginationType,
  dataLoading: PropTypes.string,
  fetchSortedItems: PropTypes.func,
  updateSortQueryParams: PropTypes.func,
};

MasterList.defaultProps = {
  data: [],
  dataSort: {
    property: 'createdAt',
    direction: 'DESC',
  },
  dataPagination: paginationDefaults,
  dataLoading: '',
  fetchSortedItems: () => {},
  updateSortQueryParams: undefined,
};
