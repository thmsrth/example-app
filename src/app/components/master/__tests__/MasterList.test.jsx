import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import staticData from '../../../static/data.json';
import MasterList from '../MasterList';

global.permissions = {
  EXAMPLE_LIST: true,
};

const { _embedded } = staticData;

const props = {
  data: _embedded.foods,
};

describe('MasterList', () => {
  it('should render the component correctly', () => {
    const tree = shallow(<MasterList {...props} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
