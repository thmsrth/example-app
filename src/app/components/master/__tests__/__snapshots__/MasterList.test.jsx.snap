// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MasterList should render the component correctly 1`] = `
<div
  className="MasterList"
>
  <div
    className="MasterList__container"
  >
    <OverviewListWrapper
      className="MasterList"
      fetchSortedItems={[Function]}
      handlers={
        Object {
          "detail": [Function],
        }
      }
      items={
        Array [
          Object {
            "calories": 266,
            "createdAt": "2018-10-28T14:48:54.01293",
            "createdBy": "thomas",
            "fat": 34,
            "id": 1,
            "name": "Pizza",
            "protein": 0.6,
            "status": "UNHEALTHY",
            "updatedAt": null,
            "updatedBy": null,
          },
          Object {
            "calories": 508,
            "createdAt": "2018-11-28T04:48:54.01293",
            "createdBy": "tommy",
            "fat": 1,
            "id": 2,
            "name": "Sushi",
            "protein": 26.7,
            "status": "HEALTHY",
            "updatedAt": null,
            "updatedBy": null,
          },
          Object {
            "calories": 271,
            "createdAt": "2018-12-28T14:48:54.01293",
            "createdBy": "ben",
            "fat": 1.3,
            "id": 3,
            "name": "Steak",
            "protein": 6.6,
            "status": "UNHEALTHY",
            "updatedAt": "2018-12-29T12:48:54.01293",
            "updatedBy": "thomas",
          },
        ]
      }
      keys={
        Array [
          Object {
            "name": "id",
            "title": "ID",
            "type": "number",
          },
          Object {
            "name": "name",
            "title": "Name",
            "type": "text",
          },
          Object {
            "name": "nutritionalInformation",
            "options": Array [
              Object {
                "name": "fat",
                "title": "Fat:",
                "type": "number",
              },
              Object {
                "name": "protein",
                "title": "Protein:",
                "type": "number",
              },
              Object {
                "name": "calories",
                "title": "Calories:",
                "type": "number",
              },
            ],
            "title": "Nutritional information",
            "type": "box",
          },
          Object {
            "name": "status",
            "title": "Status",
            "type": "text",
          },
          Object {
            "format": "DD/MM/YYYY HH:mm",
            "name": "createdAt",
            "title": "Created At",
            "type": "datetime",
          },
          Object {
            "name": "createdBy",
            "title": "Created by",
            "type": "text",
          },
          Object {
            "format": "DD/MM/YYYY HH:mm",
            "name": "updatedAt",
            "title": "Updated At",
            "type": "datetime",
          },
          Object {
            "name": "updatedBy",
            "title": "Updated by",
            "type": "text",
          },
          Object {
            "name": "actions",
            "options": Array [
              Object {
                "actionParams": Array [
                  "id",
                ],
                "condition": false,
                "icon": "search",
                "name": "detail",
                "title": "",
                "type": "action",
              },
            ],
            "title": "Actions",
            "type": "box",
          },
        ]
      }
      serverSidedSorting={false}
      sortOrder="DESC"
      sortedKey="createdAt"
    />
  </div>
</div>
`;
